bigtable/integration_test.go

```diff
@@ -2142,6 +2152,80 @@ func TestIntegration_AdminBackup(t *testing.T) {
 	}
 }
 
+// TestIntegration_DirectPathFallback tests the CFE fallback when the directpath net is blackholed.
+func TestIntegration_DirectPathFallback(t *testing.T) {
+	ctx := context.Background()
+	testEnv, _, _, table, _, cleanup, err := setupIntegration(ctx, t)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cleanup()
+
+	if !testEnv.Config().AttemptDirectPath {
+		return
+	}
+
+	if len(blackholeDpv6Cmd) == 0 {
+		t.Fatal("-it.blackhole-dpv6-cmd unset")
+	}
+	if len(blackholeDpv4Cmd) == 0 {
+		t.Fatal("-it.blackhole-dpv4-cmd unset")
+	}
+	if len(allowDpv6Cmd) == 0 {
+		t.Fatal("-it.allowdpv6-cmd unset")
+	}
+	if len(allowDpv4Cmd) == 0 {
+		t.Fatal("-it.allowdpv4-cmd unset")
+	}
+
+	if err := populatePresidentsGraph(table); err != nil {
+		t.Fatal(err)
+	}
+
+	// Precondition: wait for DirectPath to connect.
+	countEnough := examineTraffic(ctx, testEnv, table /*blackholeDP = */, false)
+	if !countEnough {
+		t.Fatalf("Failed to observe RPCs over DirectPath")
+	}
+
+	// Enable the blackhole, which will prevent communication with grpclb and thus DirectPath.
+	blackholeOrAllowDirectPath(testEnv, t /*blackholeDP = */, true)
+	countEnough = examineTraffic(ctx, testEnv, table /*blackholeDP = */, true)
+	if !countEnough {
+		t.Fatalf("Failed to fallback to CFE after blackhole DirectPath")
+	}
+
+	// Disable the blackhole, and client should use DirectPath again.
+	blackholeOrAllowDirectPath(testEnv, t /*blackholeDP = */, false)
+	countEnough = examineTraffic(ctx, testEnv, table /*blackholeDP = */, false)
+	if !countEnough {
+		t.Fatalf("Failed to fallback to CFE after blackhole DirectPath")
+	}
+}
+
+// examineTraffic counts RPCs use DirectPath or CFE traffic.
+func examineTraffic(ctx context.Context, testEnv IntegrationEnv, table *Table, expectDP bool) bool {
+	var numCount uint64
+	const (
+		numRPCsToSend  = 20
+		minCompleteRPC = 40
+	)
+
+	countEnough := false
+	start := time.Now()
+	for !countEnough && time.Since(start) < 2*time.Minute {
+		for i := 0; i < numRPCsToSend; i++ {
+			_, _ = table.ReadRow(ctx, "jadams")
+			if _, useDP := isDirectPathRemoteAddress(testEnv); useDP != expectDP {
+				numCount++
+			}
```

Comment:
Does `useDP` ever change in this loop? It appears that it will always be the same value?

Also, why send any RPCs if this will never be true? Could we just return before doing anything?

Perhaps the caller should check `isDirectPathRemoteAddress` then this function doesn't need the `expectDP` argument?

Reply:
`useDP` does get changed. `useDP` is returned by `isDirectPathRemoteAddress()`, which will return `true` if the traffic is DirectPath, and `false` if the traffic is CFE. So when DP is allowed, we expect `useDP` to be `true`, but when DP is blackholed, we expect `useDP` to be false. 

Sorry the name of the variable `expectDP` is misleading, I have rename it as `blackholeDP`. Thus, when `useDP != blackholeDP`, we get the traffic as expected. Specifically as described in the PR description, this test has three stages, as follows:

1. First wait until DirectPath traffic is observed;
In this stage, `useDP` should be `true`, and `blackholeDP` should be false.

2. Blackhole DirectPath net, and wait until the client fallback to use CFE;
In this stage, `useDP` should be `false`, and `blackholeDP` should be true.

3. Unblackhole DirectPath net, and the client should upgrade back to use DirectPath again.
The DirectPath and CFE traffic are distinguished by peer IP.
In this stage, `useDP` should be `true`, and `blackholeDP` should be false.